torrent/tracker.py

```diff
@@ -0,0 +1,378 @@
+        return action, [parse_error_to_code(error_string)]
+    else:
+        logger.warning(f"response code from server '{code}'")
+        return None 
+        
+    
+def parse_connect_packet(packet):
+    if len(packet) != 12:
+        logger.info(f"packet length for connection response not match got : {packet}")
+        return None
+    try:
+        req_response = struct.unpack(">LQ", packet)
+    except struct.debug: 
+        logger.info("parse response failed", exc_info=True)
+        return None 
+    return req_response
+
+
+def parse_announce_packet(packet):
+    if len(packet) != 16:
+        logger.info(f"packet length for announce response not match got : {packet}")
+        return None
+    try:
+        return struct.unpack(">IIII", packet)
+    except struct.error:
+        logger.info(f"parse_announce_res got from udp tracker bed response {res}")
+        return None
+
+
+def create_connect_req_struct(protocol_id=0X41727101980, transcation_id=randrange(0, 65535), action=0):
+    #the struct for udp tracker request
+    #protocol_id is const, transcation id  is random num , action 0 = connect
+    try:
+        req_struct = struct.pack(">QLL", protocol_id, action, transcation_id)
+    except struct.error:
+        raise Excpetion("the udp req struct creation failed")
+    return req_struct
+
+
+def announcing_req_packet(torrent_file, connection_id, left=0, downloaded=0, uploaded=0, event=0):
+    action = 1
+    transcation_id = randrange(0, 65535)
+    peer_id = torrent_file.peer_id.encode('utf-8')
+    ip = 0
+    key = 0
+    num_want = -1
+    port = 9999
+    try:
+        return struct.pack(">QLL20s20sQQQLLLlI", connection_id, action, transcation_id, torrent_file.info_hash, peer_id, downloaded, left, uploaded, event, ip, key, num_want, port)
+    except struct.error:
+        logger.info("creating packet failed", exc_info=True)
+        return None
+    
+     
+def get_peers(peers):
+    peer_list = []
+    for peer_pointer in range(0, len(peers), 6):
+        try:
+            peer = struct.unpack(">IH", peers[peer_pointer: peer_pointer + 6])
+        except struct.error:
+            logger.info("somthing get worng with unpack peers data")
+            return peer_list 
+        peer_list.append(peer)
+    return peer_list 
+    
+def parse_error_to_code(error):
+    """
+    1 unknown option type - info hash is not available
+    """
+    if "unknown option type" in error.lower().strip():
+        return UdpError.info_hash
+    else:
+        logger.error(f"unknown error - {error}")
+        UdpError.unknown_error
+
+#http functions
+@run_async
+def http_request(url, peer_manager,wait=0, recursive=False):
+    if wait > 1800:
+        wait = 1800
+    sleep(wait) 
+    while True:
+        logger.debug(f"start http tracker request")
+        res = None
+        try: 
+            res = requests.get(url)
+        except requests.exceptions.ConnectionError:
+            logger.debug("connect to tracker failed")
```

Comment:
maybe you should work with a helper function that gets a function, exception and log message and does this whole thing for you